packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -43,35 +41,60 @@ export default function astro({ settings, logger }: AstroPluginOptions): vite.Pl
 		name: 'astro:build',
 		enforce: 'pre', // run transforms before other plugins can
-		configResolved(_resolvedConfig) {
-			resolvedConfig = _resolvedConfig;
+		configResolved(viteConfig) {
+			// Initialize `compile` function to simplify usage later
+			compile = (code, filename) => {
+				return compileAstro({
+					compileProps: {
+						astroConfig: config,
+						viteConfig,
+						preferences: settings.preferences,
+						filename,
+						source: code,
+					},
+					astroFileToCompileMetadata,
+					logger,
+				});
+			};
 		},
 		configureServer(_server) {
 			server = _server;
+			// Make sure deleted files are removed from the compile metadata to save memory
+			server.watcher.on('unlink', (filename) => {
+				astroFileToCompileMetadata.delete(filename);
+			});
 		},
 		buildStart() {
 			astroFileToCssAstroDeps = new Map();
+			astroFileToCompileMetadata = new Map();
+
+			// Share the `astroFileToCompileMetadata` across the same Astro config as Astro performs
+			// multiple builds and its hoisted scripts analyzer requires the compile metadata from
+			// previous builds. Ideally this should not be needed when we refactor hoisted scripts analysis.
+			if (astroFileToCompileMetadataWeakMap.has(config)) {
+				astroFileToCompileMetadata = astroFileToCompileMetadataWeakMap.get(config)!;
+			} else {
+				astroFileToCompileMetadataWeakMap.set(config, astroFileToCompileMetadata);
+			}
 		},
 		async load(id, opts) {
 			const parsedId = parseAstroRequest(id);
 			const query = parsedId.query;
 			if (!query.astro) {
 				return null;
 			}
-			// For CSS / hoisted scripts, the main Astro module should already be cached
-			const filename = normalizePath(normalizeFilename(parsedId.filename, config.root));
-			let compileResult = getCachedCompileResult(config, filename);
-			if (!compileResult) {
-				// In dev, HMR could cause this compile result to be empty, try to load it first
-				if (server) {
-					await server.transformRequest('/@fs' + filename);
-					compileResult = getCachedCompileResult(config, filename);
-				}
 
-				// If there's really no compilation result, error
-				if (!compileResult) {
-					throw new Error('No cached compile result found for ' + id);
-				}
+			// Astro scripts and styles virtual module code comes from the main Astro compilation
+			// through the metadata from `astroFileToCompileMetadata`. It should always exist as Astro
+			// modules are compiled first, then its virtual modules. If the virtual modules are somehow
+			// compiled first, throw an error and we should investigate it.
+			const filename = normalizePath(normalizeFilename(parsedId.filename, config.root));
+			const compileMetadata = astroFileToCompileMetadata.get(filename);
+			if (!compileMetadata) {
+				throw new Error(
+					`No cached compile metadata found for "${id}". The main Astro module "${filename}" should have ` +
+						`compiled and filled the metadata first, before its virtual modules can be requested.`
+				);
```

Comment:
We should always use the AstroError subclass if we can. I think this would fit under `UnknownCompilerError`?

Reply:
I used an error here since the other ones are also an error 😬 I think they are more of unexpected errors as our normal code paths should never trigger the error in the first place, so `UnknownCompilerError` might not be the most accurate.

I could switch all the errors to a new Astro error in a separate PR though so it's more consistent. Hoping to keep the PR scope down a bit!